Dev

VincentHuet opened the pull request:
export opérationnel

Commit:
correction suivant les recommandations:
- manque cancan
- devons disctuter vis-à-vis de primary_locale

Commit:
mise en place de cancan

Commit:
renommage de certaines variables
amélioration de la lisibilité des status des traductions
ajustement du fonctionnement des droits de modification d'une translation
correction de bug lié à la modification du modeèle "locale"

Commit:
modification mineure

Commit:
correction sur les helpers

Commit:
modification des controller

Commit:
remodification des controleurs

Commit:
correction mais beug de chargement

Commit:
modification site presque opérationnel

Commit:
modification de "translations_controlleur": appli => OK
manque export xml

Commit:
remise à plat des commits/pull/etc..

Commit:
modification des controlleurs

Commit:
modification de la fonction create dans les controleurs "translat.."

Commit:
modification mineure

Commit:
modification de "translation_controlleur" appli_ok

Commit:
simplification de "application_controler"

Commit:
modification des "controlleurs" pour les "class" Yml_sources et Phrases

Commit:
modification de "locale_controller"

Commit:
mise en place des premières pistes pour l'export

Commit:
mise en place de la création d'un hash lors de la consultation d'une page de "translation"

Commit:
correctif pour améliorer le ranking des classes

Commit:
Ajustement de la construction du hash
Mise en place de l'export
correction pour un affichage correct vis à vis de la synthaxe YAML
mise à jour de chaque locale située dans public/data des que l'on consulte les traductions correspondantes

Commit:
modification pour corriger le ranking de la classe translation helper

djeusette commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,54 @@
+  protected
+  def standard_respond_to(class_type)
+    respond_to do |format|
+      format.html # show.html.erb
```
remove the comment

djeusette commented on app/controllers/translations_controller.rb:
```diff
@@ -1,24 +1,33 @@
-    respond_to do |format|
-      format.html # index.html.erb
-      format.json { render json: @translations }
+    if !Locale.find(params[:locale_id]).nil?
```
should raise an error if no locale found.
i.e. ActiveRecord::RecordNotFound

Use Locale.find_by_id(params[:locale_id]) instead

@locale = Locale.find_by_id(params[:locale_id])
if @locale 
  @translations = @locale.translations (no need for the .all)
else
  @translations = Translation.all
end

djeusette replied:
@locale and not @Locale

(sorry ;))

djeusette replied:
Holy shit !!!

no maj at the word locale for initializing the variable...

djeusette commented on app/controllers/translations_controller.rb:
```diff
@@ -1,24 +1,33 @@
+      @translations = Translation.all
     end
+    
+    @yml_hash = {}
```
extract the block to init @yml_hash to a private method

djeusette commented on app/helpers/locales_helper.rb:
```diff
@@ -1,18 +1,17 @@
 module LocalesHelper
 
   def primary(locale)
```
ternary operator ?

locale.primary_locale ? "primary" : "secondary"

djeusette commented on app/helpers/translators_helper.rb:
```diff
@@ -1,2 +1,6 @@
 module TranslatorsHelper
+
+  def locale_list
```
We're in a helper.

It's a smell to access the DB via a helper.
Instead do this:

def locale_list(locales)
  locales.collect {|l| [l.name, l.id]}
end

and before calling this helper method, just initialize a global variable in the controller
@locales (in the controller)

then call the helper method like this:

locale_list(@locales)

djeusette commented on app/models/translation.rb:
```diff
@@ -21,14 +21,10 @@ class Translation < ActiveRecord::Base
-  def need_updated?(current_translator)
-    self.needed_update == 1 && current_translator.locale_id == self.locale_id || Locale.find(self.locale_id).is_primary? && Locale.find(current_translator.locale_id).is_primary? 
+  def need_updated?
+    self.needed_update == 1 
```
no need for the self.

djeusette commented on app/views/blog/index.html.haml:
```diff
@@ -1,11 +1,11 @@
-    = current_translator.id
-  %br
-    = Locale.find(current_translator.locale_id).name
+/ - if translator_signed_in?
```
why commented ?

djeusette commented on app/views/devise/registrations/new.html.haml:
```diff
@@ -10,7 +10,7 @@
   .field
     = f.label :locale_id
-    = f.select(:locale_id, Locale.all.collect {|l| [ l.name,l.id ] }, { :include_blank => false })
+    = f.select(:locale_id, locale_list, { :include_blank => false })
```
locale_list(@locales) as said before

Commit:
Asked modifications : done

Commit:
correctif demandé par David
mise en place d'un algo permettant de passer à une nouvelle traduction une fois la présente validée

Commit:
modification du système d'affichage: diagnostique : porblème au niveau du "yaml_loader"

Commit:
Correctif sur la construction des "yaml_path" et leur ré-interprétation lors de l'export

Commit:
modification de l'architecture
- mise en place d'un nouveau domain
- création d'un bouton pour exporter en yaml
- ré-organisation du code pour etre plus restfull/crud au niveau du chargement en DB

Commit:
essaye de modification pour mise à jour "in line"

Commit:
modification avec marc

Commit:
mise en place du in line avec David

Commit:
modification user friendly de l'accueil
modification générale

Commit:
Correction des bugs mineurs
Facelift du site => become userfriendly

Commit:
mise en place d'icon et autre style

Commit:
Modification des couleurs et ajout d'icone
Redirection vers le bon répertoire des fichiers yml

Commit:
Refactoring du controleur des "translations"
et ajustement mineur

Commit:
refactor de "translation_in_line" et modif mineure

Commit:
Correction bug d'affichage dans la page "récap..."
Manque refactoring dès que la connexion avec code climate sera ok

Commit:
debut des modifications demandées par Ani

Commit:
Modification suggérée par Ani => done

Commit:
modification pour rendre de nouveau opérationel l'export + historisation

Commit:
nouvelle modification pour garantir le bon fonctionnement de l'application

Commit:
modif avec dav

Commit:
Fin des modification demander par David : translation in line pour admin + style
Debut des modifs demandées par Tanguy

Commit:
modification démandées pas Tanguy effectuée
manque inject et textarea elastic

Commit:
modification pour obtenir une "elastic textarea"
modification de l'estétique et ajustement des menu en fonction de la catégorie d'utilisateur

Commit:
mis en place du inject
reste la gestion des values modifié

Commit:
mise en place des modifications reglant le problème de la mise à jour des "translations" primaires si la valeur "text" est modifiée

djeusette commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,131 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery
+  before_filter :initialise_locales_name
```
initialize (spelling mistake)

djeusette commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,131 @@
+
+  rescue_from CanCan::AccessDenied do |exception|
+    flash[:error] = "Access denied."
+    redirect_to root_url
```
if the user was not logged in, that could be nice to redirect him to the page he wanted to access before being redirected.

try something like
==>

if !user_signed_in?
   session[:translator_return_to] = request.url
   redirect_to new_translator_session_path, :alert => exception.message
end

djeusette replied:
of course you need an else statement to handle the case where the user was logged in but can not access the page.

for you to understand a bit, after a user (a translator in this case) has logged in, the method "after_sign_in_path_for(resource)" from devise is called.  It checks if a path is stored somewhere and redirect to the latter or to a default path if not found.

Here, we put the after sign in path in session.

Not sure it will work as-is but worth trying.

djeusette replied:
Here is the source code of stored_location_for(resource)

def stored_location_for(resource_or_scope)
  scope = Devise::Mapping.find_scope!(resource_or_scope)
  session.delete("#{scope}_return_to")
end

djeusette commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,131 @@
+  def create_phrase_translation_table(translations)
+    relevant_phrase_text = {}
+    translations.each do |translation|
+      primary_locale = Locale.primary_locale
```
Those two lines can be extracted from the each block to avoid unnecessary DB queries

(primary_locale and first_locale_translations)

djeusette commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,131 @@
+      primary_locale = Locale.primary_locale
+      first_locale_translations = primary_locale.translations
+      relevant_phrase = first_locale_translations.where(:phrase_id => translation.phrase_id)
+      relevant_phrase_text[translation.phrase_id] = relevant_phrase.first.text
```
relevant_phrase should be plural as this returns an ActiveRecord::Collection => relevant_phrases

Then, for relevant_phrase_text you don't check if relevant_phrases.first is null or not.  In case it equals nil, nil.text will raise an error.

djeusette replied:
we end up with the following inject ;)

primary_locale = Locale.primary_locale
first_locale_translations = primary_locale.translations

translations.inject({}) do |relevant_phrase_text, translation|
  relevant_phrases = first_locale_translations.where(:phrase_id => translation.phrase_id)
  relevant_phrase_text[translation.phrase_id] = relevant_phrases.first.text if relevant_phrases.any?
end

djeusette commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,131 @@
+    primary_keyword_quantity - current_done_translation
+  end
+
+  def phrase_quantity()
```
just remove the brackets when no args

djeusette commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,131 @@
+  end
+
+  def missing_translation?(locale)
+    primary_keyword_quantity = phrase_quantity
```
can remove this first line and change primary_keyword_quantity by phrase_quantity

djeusette commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,131 @@
+  def phrase_quantity()
+    reference_locale = Locale.primary_locale
+    primary_keyword_quantity = reference_locale.translations.count
+    primary_keyword_quantity
```
This last line is not required as the assignment of the previous line is already a function (i.e. it returns a value)

djeusette commented on app/controllers/blog_controller.rb:
```diff
@@ -5,9 +5,21 @@ class BlogController < ApplicationController
   require 'yaml'
 
   def index
```
blog_controller ?

what's the point ?  to remove ?

djeusette replied:
anyway: needs permissions or the guest user case (user not signed in) should be handled

VincentHuet replied:
due to "load_and_authorize_resource ", I get an error because : 

" 
NameError in HomeController#index
uninitialized constant Home
"

djeusette commented on app/controllers/locales_controller.rb:
```diff
@@ -2,43 +2,34 @@
+  load_and_authorize_resource
 
   def index
     @locales = Locale.all
```
can be removed due to the load_and_authorize_resource

djeusette commented on app/controllers/locales_controller.rb:
```diff
@@ -2,43 +2,34 @@
   # GET /locales/new
   # GET /locales/new.json
   def new
     @locale = Locale.new
```
can be removed

djeusette commented on app/controllers/locales_controller.rb:
```diff
@@ -46,16 +37,7 @@ def edit
   def create
     @locale = Locale.new(params[:locale])
```
can be removed

djeusette commented on app/controllers/locales_controller.rb:
```diff
@@ -63,15 +45,7 @@ def create
   def update
     @locale = Locale.find(params[:id])
```
remove

VincentHuet replied:
remove!remove!remove!remove!remove!!!!!remove!!!!!removeremove!!remove

djeusette commented on app/controllers/locales_controller.rb:
```diff
@@ -80,9 +54,13 @@ def destroy
     @locale = Locale.find(params[:id])
```
remove!

djeusette commented on app/controllers/missing_translations_controller.rb:
```diff
@@ -0,0 +1,8 @@
+class MissingTranslationsController < ApplicationController
+
```
permissions ?

authorize! :read, @missing_translations if in the index method

this will use the :read permission of Translation (check in ability.rb)

djeusette commented on app/controllers/phrases_controller.rb:
```diff
@@ -4,41 +4,28 @@ class PhrasesController < ApplicationController
+  load_and_authorize_resource
 
   def index
     @phrases = Phrase.all
```
remove

djeusette commented on app/controllers/phrases_controller.rb:
```diff
@@ -4,41 +4,28 @@ class PhrasesController < ApplicationController
   # GET /phrases/new
   # GET /phrases/new.json
   def new
     @phrase = Phrase.new
```
remove

djeusette commented on app/controllers/phrases_controller.rb:
```diff
@@ -46,43 +33,21 @@ def edit
   def create
     @phrase = Phrase.new(params[:phrase])
```
remove

djeusette commented on app/controllers/translations_controller.rb:
```diff
@@ -1,23 +1,45 @@
-    @translations = @locale.translations.all
+    @locale = Locale.find_by_id(params[:locale_id])
+    if @locale
+      @translations = @locale.translations.order('text')
```
Ok here because cancan only load the resources if it is not already initialized.
In this case, the initialization is not trivial, so you need to do it yourself.

djeusette commented on app/controllers/translations_controller.rb:
```diff
@@ -1,23 +1,45 @@
+      @latest_translation_load_date = Translation.maximum("created_at").to_date
     end
+
+    @create_date = {}
```
remove this line (the hash is init in the create...)

djeusette commented on app/controllers/translations_controller.rb:
```diff
@@ -1,23 +1,45 @@
+    @create_date = {}
+    @create_date = create_create_date_table(@translations)
+
+    @update_date = {}
```
remove

djeusette commented on app/controllers/translations_controller.rb:
```diff
@@ -48,60 +66,67 @@ def edit
+  def create_update_date_table(translations)
+    update_date = {}
+    translations.each do |translation|
+      update_date[translation.updated_at.to_date] = translation.updated_at.to_date
```
????

What's the point ?

djeusette commented on app/controllers/translations_controller.rb:
```diff
@@ -48,60 +66,67 @@ def edit
+    create_date = {}
+    translations.each do |translation|
+      if translation.need_updated? && translation.created_at.to_date != @latest_translation_load_date
+        create_date[translation.created_at.to_date] = translation.created_at.to_date
```
????

djeusette replied:
keys and values are the same

djeusette replied:
otherwise, this is perfect for an inject :D

djeusette commented on app/controllers/translations_in_line_controller.rb:
```diff
@@ -0,0 +1,36 @@
+class TranslationsInLineController < ApplicationController
```
permissions

djeusette commented on app/controllers/translators_controller.rb:
```diff
@@ -4,84 +4,53 @@ class TranslatorsController < ApplicationController
+  load_and_authorize_resource
+
   def index
     @translators = Translator.all
```
remove

djeusette commented on app/controllers/translators_controller.rb:
```diff
@@ -4,84 +4,53 @@ class TranslatorsController < ApplicationController
   # GET /translators/new
   # GET /translators/new.json
   def new
     @translator = Translator.new
```
remove

djeusette commented on app/controllers/translators_controller.rb:
```diff
@@ -4,84 +4,53 @@ class TranslatorsController < ApplicationController
   # POST /translators
   # POST /translators.json
   def create
     @translator = Translator.new(params[:translator])
```
remove

djeusette commented on app/controllers/yml_files_controller.rb:
```diff
@@ -0,0 +1,24 @@
+class YmlFilesController < ActionController::Base
+
```
permissions

djeusette commented on app/controllers/yml_sources_controller.rb:
```diff
@@ -1,35 +1,30 @@
   # GET /yml_sources.json
 
   before_filter :authenticate_translator!
+  #load_and_authorize_resource
```
why commented ?

if the permissions are well established, you don't need the before_filter (authenticate_translator)

djeusette commented on app/controllers/yml_sources_controller.rb:
```diff
@@ -1,35 +1,30 @@
   # GET /yml_sources/1
   # GET /yml_sources/1.json
   def show
-    @yml_source = YmlSource.find(params[:id])
```
will not work without load_and_authorize_resource

djeusette commented on app/domain/translations_manager.rb:
```diff
@@ -0,0 +1,57 @@
+
+    head, *tail = path
+    if tail.empty?
+      puts "----------!!!--- if ---!!!------------"
```
remove all puts if it works properly

djeusette commented on app/domain/translations_manager.rb:
```diff
@@ -0,0 +1,57 @@
+
+    translations.each do |translation|
+      translation_phrase = translation.phrase
+      yaml_path = translation_phrase.yaml_path
```
to avoid this kind of "ask don't tell"

use (in the Translation model)
===>

delegate :yaml_path, :to => :phrase

djeusette replied:
Then you'll be able to directly make this call:

translation.yaml_path

djeusette commented on app/domain/translations_manager.rb:
```diff
@@ -0,0 +1,57 @@
+    yml_hash
+  end
+
+  def self.hash_to_export(locale_translation_hash, indent, space_type, return_type)
```
Avoid as much as possible the use of + to concatenate two strings.

If one of the two is nil, it will raise an error.

Instead use arrays and the join method.

djeusette commented on app/mailers/translator_mailer.rb:
```diff
@@ -5,6 +5,6 @@ class TranslatorMailer < ActionMailer::Base
     @translator = translator
     @url  = "http://example.com/login"
-    mail(:to => translator.email, :subject => "Welcome to My Awesome Site")
+    mail(:to => @translator.email, :subject => "New translation are needed ! ")
```
There's a misspelling in the subject.

New translationS

djeusette commented on app/models/locale.rb:
```diff
@@ -16,14 +16,13 @@ class Locale < ActiveRecord::Base
   has_many :translations
```
you should always add the :dependent => :destroy (sometimes :nullify for logging purpose or when it is relevant to still access the data) operand to avoid cluttering the DB with never accessed entries/rows.

Commit:
simplification de l'index de translation
correction de la construction du yaml_path dans yml_loader

Commit:
modification demandé par Tanguy&David

Commit:
Fin des modifications demandées par David & Tanguy
modification des fonctions d'output => intégrité du yml en sortie
